Enable configuring test reporters via config

gordonsyme opened the pull request:
Fixes #4 

Use `load-reader` to load the whole config file, this allows config code to
import namespaces sensibly.
Note: namespaces could always be imported by invoking `require` in a `do` or
`let` form inside the config map definition so this shouldn't introduce any
additional risk.

`test-ns`/`test-var` make sure that config is only read once per test
namespace, this is important to retain the same test reporter objects for all
tests belonging to the test namespace.

Add sample config that instantiates test reporters

I still need to work out how to test this properly. I've confirmed it works by
updating another repo to use this circleci.test code and setting a config.clj.

I am still concerned about looking up a resource and loading the code within it
unconditionally, some additional checks to ensure it came from the current
project are necessary.

gordonsyme committed:
Enable configuring test reporters via config

Use `load-reader` to load the whole config file, this allows config code
to import namespaces sensibly.
Note: namespaces could always be imported by invoking `require` in a
`do` or `let` form inside the config map definition so this shouldn't
introduce any additional risk.

test-ns/test-var make sure that config is only read once per test
namespace, this is important to retain the same test reporter objects
for all tests belonging to the test namespace.

gordonsyme committed:
Add sample config that instantiates test reporters

technomancy reviewed (commented):
My only minor hesitation here is that we can now have `read-config` called in more than one place for a single test run. (When invoked from `-main`.)

What if we deferred the lookup of the selector (currently happening in `read-args`) up to `test-ns` so that the config reading could all be done in one place?

gordonsyme commented:
Works for me, I'll make that happen.

gordonsyme committed:
Only read config once regardless of entry-point

Rename `read-config` to `read-config!`

Make `run-selected-tests` private to avoid needing multiple varargs
arities (which is not allowed)

technomancy commented on src/circleci/test.clj:
```diff
@@ -110,9 +113,11 @@
   ([ns] (test-ns ns (constantly true)))
-  ([ns selector]
+  ([ns selector] (test-ns (read-config!) ns selector))
+  ([config ns selector]
```
It's a bit unusual to have an optional arg at the front of the arglist; `config` here should probably get moved to the end.

gordonsyme commented:
Closing for the moment, this isn't where I want it to be and I can't dedicate the time right now.

technomancy commented:
I thought this was more or less ready to merge other than the minor argument list issue above. Can you explain what else you think is wrong with it?

gordonsyme commented:
I think the config reading could happen in a better place, that and I don't have tests written. I can't work on it until July at the earliest so felt better to close than leave it languish.

Feel free to pick it up!
